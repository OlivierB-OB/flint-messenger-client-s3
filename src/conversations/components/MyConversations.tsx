import Box from '@material-ui/core/Box';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
import React from 'react';
import { connect } from 'react-redux';
import { Link, useRouteMatch } from 'react-router-dom';
import { IAppState } from '../../appReducer';
import { IConversation, IConversationsStatus } from '../types';
import { ConversationListItem } from './ConversationListItem';
import { Loading } from '../../layout/components/Loading';

export interface IConversationListProps {
  status: IConversationsStatus;
  list: IConversation[];
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    selected: {
      color: theme.palette.secondary.contrastText,
      backgroundColor: theme.palette.secondary.light,
    },
  }),
);

export function ConversationList({ status, list }: IConversationListProps) {
  const loading = status === 'unavailable' ? <Loading /> : null;
  const { selected } = useStyles();
  const match = useRouteMatch<{ conversationId: string }>('/conversation/:conversationId');
  const conversationId = match?.params.conversationId;
  return (
    <Box style={{ minWidth: '300px' }}>
      {loading}
      <List>
        {loading || list.length ? null : (
          <ListItem>
            <ListItemText primary="no conversation available..." />
          </ListItem>
        )}
        {list.map((conversation) => (
          <ListItem
            className={conversationId === conversation._id ? selected : undefined}
            button
            component={Link}
            to={`/conversation/${conversation._id}`}
            key={conversation._id}
          >
            <ConversationListItem conversation={conversation} />
          </ListItem>
        ))}
      </List>
    </Box>
  );
}

const mapStateToProps = ({ conversations }: IAppState) => ({
  status: conversations.status,
  list: conversations.conversations,
});

export const MyConversations = connect(mapStateToProps)(ConversationList);
