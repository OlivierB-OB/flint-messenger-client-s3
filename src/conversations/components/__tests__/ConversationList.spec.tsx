import React from 'react';
import { shallow } from 'enzyme';
import { ConversationList } from '../MyConversations';

describe('ConversationList', () => {
  it('should appear as loading if conversation service is unavailable', async () => {
    const props: any = {
      status: 'unavailable',
      list: [],
    };
    const component = shallow(
      <ConversationList {...props} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });

  it('should display a message if no conversation', async () => {
    const props: any = {
      status: 'ready',
      list: [],
    };
    const component = shallow(
      <ConversationList {...props} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });

  it('should display the conversations', async () => {
    const props: any = {
      status: 'unavailable',
      list: [
        { _id: '123' },
        { _id: '456' },
      ],
    };
    const component = shallow(
      <ConversationList {...props} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });
});
