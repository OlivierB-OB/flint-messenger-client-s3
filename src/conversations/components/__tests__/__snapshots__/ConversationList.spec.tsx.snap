// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationList should appear as loading if conversation service is unavailable 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <Loading />
  <WithStyles(ForwardRef(List)) />
</div>
`;

exports[`ConversationList should display a message if no conversation 1`] = `
<div
  className="MuiBox-root MuiBox-root-2"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))>
      <WithStyles(ForwardRef(ListItemText))
        primary="no conversation available..."
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</div>
`;

exports[`ConversationList should display the conversations 1`] = `
<div
  className="MuiBox-root MuiBox-root-3"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <Loading />
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      key="123"
      to="/conversation/123"
    >
      <ConversationListItem
        conversation={
          Object {
            "_id": "123",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      key="456"
      to="/conversation/456"
    >
      <ConversationListItem
        conversation={
          Object {
            "_id": "456",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</div>
`;
